pkg/planner/core/indexmerge_path.go

```diff
@@ -210,48 +214,55 @@ func (ds *DataSource) generateIndexMergeOrPaths(filters []expression.Expression)
 
 			itemPaths := ds.accessPathsForConds(pushedDownCNFItems, usedIndexCount)
 			if len(itemPaths) == 0 {
-				partialPaths = nil
+				partialAlternativePaths = nil
 				break
 			}
-			partialPath := ds.buildIndexMergePartialPath(itemPaths)
-			if partialPath == nil {
-				partialPaths = nil
-				break
+			// we don't prune other possible index merge path here.
+			// keep all the possible index merge partial paths here to let the property choose.
+			partialAlternativePaths = append(partialAlternativePaths, itemPaths)
+		}
+		// in this loop we do two things.
+		// 1: If all the partialPaths use the same index, we will not use the indexMerge.
+		// 2: Compute a theoretical best countAfterAccess(pick its accessConds) for every alternative path(s).
+		indexMap := make(map[int64]struct{}, 1)
+		accessConds := make([]expression.Expression, 0, len(partialAlternativePaths))
+		for i := len(partialAlternativePaths) - 1; i >= 0; i-- {
+			// 1: mark used map.
+			for j := len(partialAlternativePaths[i]) - 1; j >= 0; j-- {
+				if partialAlternativePaths[i][j].IsTablePath() {
+					// table path
+					indexMap[-1] = struct{}{}
+				} else {
+					// index path
+					indexMap[partialAlternativePaths[i][j].Index.ID] = struct{}{}
+				}
 			}
-			partialPaths = append(partialPaths, partialPath)
-		}
-		// If all of the partialPaths use the same index, we will not use the indexMerge.
-		singlePath := true
-		for i := len(partialPaths) - 1; i >= 1; i-- {
-			if partialPaths[i].Index != partialPaths[i-1].Index {
-				singlePath = false
-				break
+			// 2.1: trade off on countAfterAccess.
+			theoreticalMinCountAfterAccessPath := ds.buildIndexMergePartialPath(partialAlternativePaths[i])
```

Comment:
```suggestion
			minCountAfterAccessPath := ds.buildIndexMergePartialPath(partialAlternativePaths[i])
```
I think we don't need to specially say it's "theoretical".